arangod/Aql/RegisterPlan.cpp

```diff
@@ -83,6 +83,24 @@ void RegisterPlanWalkerT<T>::after(T* en) {
     }
   };
 
+  auto const updateRegsToKeep = [this](T* en, VarSet const& varsUsedLater,
+                                       VarSet const& varsValid) {
+    auto const& varsSetHere = en->getVariablesSetHere();
+
+    // regsToKeep
+    regsToKeepStack.back().clear();
+    for (auto const var : varsValid) {
+      bool isSetHere = std::find(varsSetHere.begin(), varsSetHere.end(), var) !=
+                       varsSetHere.end();
+      bool isUsedLater = std::find(varsUsedLater.begin(), varsUsedLater.end(), var) !=
+                         varsUsedLater.end();
+      if (!isSetHere && isUsedLater) {
```

Comment:
as a slight optimization, we could save the calculation of `isUsedLater` when `isSetHere` is false, but I am not sure if it is worth it. It _may_ be worth if the function is called often and finding a variable in `varsUsedLater` is expensive.